### PR TITLE
pkg/edhoc-c: initial commit

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds initial support for the [EDHOC-C](https://github.com/openwsn-berkeley/EDHOC-C) [edhoc](https://datatracker.ietf.org/doc/draft-ietf-lake-edhoc/) protocol. The pkg is up to date with draft 5. A test application is included that allows performing the handshake over COAP.

The current examples uses RPK keys, but x509 and c509 are also possible

Currently, the only supported crypto backend is wolfssl which is not very RAM friendly, the pkg currently supports HACL as well, but not the version in RIOT master.

@TimothyClaeys might want to take a look and can give a better status of the EDHOC implementation.

### Testing procedure

On native:

```
PORT=tap0 make -C tests/pkg_edhoc_c all test-with-config
```

On samr21-xpro:

```
BOARD=samr21-xpro make -C tests/pkg_edhoc_c flash test-with-config
```

```
2021-04-09 16:14:40,967 # init handshake fe80::204:2519:1801:a9ba
2021-04-09 16:14:40,970 # [initiator]: sending msg1 (37 bytes):
2021-04-09 16:14:40,974 # 0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
2021-04-09 16:14:40,978 # 0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
2021-04-09 16:14:40,982 # 0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
2021-04-09 16:14:40,986 # 0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
2021-04-09 16:14:40,988 # 0xe2 0x44 0x1a 0x7c 0x21
2021-04-09 16:14:40,993 # [responder]: received an EDHOC message (len 37):
2021-04-09 16:14:40,997 # 0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
2021-04-09 16:14:41,001 # 0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
2021-04-09 16:14:41,005 # 0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
2021-04-09 16:14:41,009 # 0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
2021-04-09 16:14:41,011 # 0xe2 0x44 0x1a 0x7c 0x21
2021-04-09 16:14:44,905 # [responder]: sending msg2 (46 bytes):
2021-04-09 16:14:44,909 # 0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
2021-04-09 16:14:44,912 # 0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
2021-04-09 16:14:44,916 # 0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
2021-04-09 16:14:44,920 # 0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
2021-04-09 16:14:44,924 # 0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
2021-04-09 16:14:44,927 # 0x02 0x8d 0x19 0xcf 0x3c 0x99
2021-04-09 16:14:44,931 # [initiator]: received a message (46 bytes):
2021-04-09 16:14:44,935 # 0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
2021-04-09 16:14:44,939 # 0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
2021-04-09 16:14:44,943 # 0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
2021-04-09 16:14:44,947 # 0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
2021-04-09 16:14:44,951 # 0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
2021-04-09 16:14:44,953 # 0x02 0x8d 0x19 0xcf 0x3c 0x99
2021-04-09 16:14:50,793 # [initiator]: sending msg3 (20 bytes):
2021-04-09 16:14:50,797 # 0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
2021-04-09 16:14:50,800 # 0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
2021-04-09 16:14:50,803 # 0xe0 0xa8 0x1b 0xbf
2021-04-09 16:14:50,808 # [responder]: received an EDHOC message (len 20):
2021-04-09 16:14:50,811 # 0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
2021-04-09 16:14:50,815 # 0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
2021-04-09 16:14:50,817 # 0xe0 0xa8 0x1b 0xbf
2021-04-09 16:14:50,820 # [responder]: finalize exchange
2021-04-09 16:14:52,770 # [responder]: handshake successfully completed
2021-04-09 16:14:52,774 # [initiator]: handshake successfully completed
2021-04-09 16:14:52,778 # [initiator]: Transcript hash 4 (32 bytes):
2021-04-09 16:14:52,782 # 0x7c 0xcf 0xde 0xdc 0x2c 0x10 0xca 0x03
2021-04-09 16:14:52,786 # 0x56 0xe9 0x57 0xb9 0xf6 0xa5 0x92 0xe0
2021-04-09 16:14:52,789 # 0xfa 0x74 0xdb 0x2a 0xb5 0x4f 0x59 0x24
2021-04-09 16:14:52,793 # 0x40 0x96 0xf9 0xa2 0xac 0x56 0xd2 0x07
2021-04-09 16:14:52,793 #
```

### CODE REVIEW

File: tests/pkg_edhoc_c/initiator.c [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+static edhoc_ctx_t _ctx;
+static edhoc_conf_t _conf;
+static rpk_t _rpk;
+static cred_id_t _cred_id;
+static cose_key_t _auth_key;
+static wc_Sha256 _wc_sha256;
+static uint8_t _method;
+static uint8_t _suite;
+
+static ssize_t _send(coap_pkt_t *pkt, size_t len, char *addr_str, uint16_t port)
```
Comment (fjmolinas) [inline tests/pkg_edhoc_c/initiator.c:51]: @kaspar030 maybe you can take a look at this section?

Comment (kaspar030) [reply tests/pkg_edhoc_c/initiator.c:51]: I seems ok. Is there a reason why you're not using `sock_udp_str2ep()`?

... Looking into that, I see that your function does actually verify whether the netif is a valid gnrc interface, so this cool. We should make this functionality available to sock at some point.

File: pkg/edhoc-c/doc.txt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * And don't forget to include the header for the EDHOC-C public API:
+ *
+ * ```c
+ * #include <edhoc/edhoc.h>
+ * ```
+ *
+ * ### Current Support
+ *
+ * - EDHOC-C support @ref pkg_hacl as a crypto backend but its using a different
```
Comment (kaspar030) [inline pkg/edhoc-c/doc.txt:44]:
```suggestion
 * - EDHOC-C supports @ref pkg_hacl as crypto backend but it's using a different
```

File: pkg/edhoc-c/doc.txt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ *   version than the one supported currently in RIOT
+ * - x509 certificates required MBED-TLS, which is currently not supported in
```
Comment (kaspar030) [inline pkg/edhoc-c/doc.txt:46]:
```suggestion
 * - x509 certificates require MBED-TLS, which is currently not supported in
```

File: pkg/edhoc-c/doc.txt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ *   RIOT so only RPK and CBOR certificates are possible.
```
Comment (kaspar030) [inline pkg/edhoc-c/doc.txt:47]:
```suggestion
 *   RIOT, so only RPK and CBOR certificates are supported.
```

File: tests/pkg_edhoc_c/common.c [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+#include <stdio.h>
+#include <string.h>
+
+#include "edhoc/edhoc.h"
+#include "edhoc_keys.h"
+
+#define ENABLE_DEBUG        0
+#include "debug.h"
+
+#define CRED_DB_SIZE        ARRAY_SIZE(cred_db)
```
Comment (kaspar030) [inline tests/pkg_edhoc_c/common.c:29]: ARRAY_SIZE *should* need an include

File: tests/pkg_edhoc_c_unittest/Makefile [hunk 6]
```diff
@@ -0,0 +1,13 @@
+include ../Makefile.tests_common
```
Comment (kaspar030) [inline tests/pkg_edhoc_c_unittest/Makefile:1]: IMO this application can be dropped?

File: tests/pkg_edhoc_c/common.c (hunk 5 above)
Comment (fjmolinas) [reply tests/pkg_edhoc_c/common.c:30]: addressed

**Comment by fjmolinas** [comment]
Will trigger a build to find issues.

**Comment by kaspar030** [comment]
btw, please give this a squash when the build issues are fixed, I'll give it another reviewing round then

**Comment by kaspar030** [comment]
btw, please give this a squash when the build issues are fixed, I'll give it another reviewing round then

**Comment by fjmolinas** [comment]
> btw, please give this a squash when the build issues are fixed, I'll give it another reviewing round then

squashed

File: tests/pkg_edhoc_c/README.md [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+INFO:root:EDHOC key exchange successfully completed:
+INFO:root: - connection IDr: b'+'
+INFO:root: - connection IDi: b'+'
+INFO:root: - aead algorithm: AES_CCM_16_64_128
+INFO:root: - hash algorithm: SHA_256
+INFO:root: - OSCORE secret : b'\xd8\x1e\xa3@\xec\xe3?3\xe1\xfe\x8a\x1d\x0c|\xd0\xbe'
+INFO:root: - OSCORE salt   : b'\x87\xf9J\xf7\x82Tq\xa3'
+```
+
+Congratulations you have performed an EDHOC handshake form your HOST to
```
Comment (kaspar030) [inline tests/pkg_edhoc_c/README.md:226]: from

File: tests/pkg_edhoc_c/README.md [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+0xb5 0x0c 0x32 0x31 0x9f 0x34 0x0f 0x4a
+0xc5 0xaf 0x9a
+
+[responder]: finalize exchange
+[responder]: handshake successfully completed
+```
+
+### Initiator
+
+First find out what is the local ipv6 address of the tap interface:
```
Comment (kaspar030) [inline tests/pkg_edhoc_c/README.md:137]:
```suggestion
First find out the local ipv6 address of the tap interface:
```

File: tests/pkg_edhoc_c/README.md [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+This will create a tap interface called `riot0`, owned by the user. It will
+also run an instance of uhcpcd, which starts serving the prefix
+`2001:db8::/64`. Keep the shell open as long as you need the network.
+Make sure to exit the "make term" instance from the next section *before*
+exiting this, as otherwise the "riot0" interface doesn't get cleaned up
+properly.
+
+### Responder
+
+Find out what the IPv6 address of the device is by running the `ifconfig`
```
Comment (kaspar030) [inline tests/pkg_edhoc_c/README.md:52]:
```suggestion
Find out the IPv6 address of the device by running the `ifconfig`
```

File: pkg/edhoc-c/Makefile.dep [hunk 10]
```diff
@@ -0,0 +1,22 @@
+ifneq (,$(filter edhoc-c_crypto_wolfssl,$(USEMODULE)))
```
Comment (kaspar030) [inline pkg/edhoc-c/Makefile.dep:1]: there is no default for crypto and cbor backend, right? (let's fix later if necessary)

**Review by kaspar030 — CHANGES_REQUESTED** [review]
last round, there are some typos. please squash right in

File: tests/pkg_edhoc_c/tests-with-config/01-run.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+import os
+import sys
+
+from testrunner import run
+
+# Default COAP port on which the response server is running
+COAP_PORT = int(os.getenv("COAP_PORT", "5683"))
+
+LAKE_WG_EDHOC_TV_34900_MSG1 = \
```
Comment (fjmolinas) [inline tests/pkg_edhoc_c/tests-with-config/01-run.py:17]: By addressing flake9 test does not pass, will fix

File: tests/pkg_edhoc_c/tests-with-config/01-run.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+LAKE_WG_EDHOC_TV_34900_OSCORE_SALT = \
+    "0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d"
+
+
+def get_ipv6_addr(child):
+    child.expect_exact('>')
+    child.sendline('ifconfig')
+    # Get device local address
+    child.expect(
+        r"inet6\s+addr:\s+(?P<lladdr>[0-9a-fA-F:]+:[A-Fa-f:0-9]+)"
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply tests/pkg_edhoc_c/tests-with-config/01-run.py]: sorry, this was correct before, I didn't realize that the regexp spans two lines. this is broken now, so please revert this suggestion

**Comment by kaspar030** [comment]
please squash!

**Review by kaspar030 — APPROVED** [review]
ACK.

**Comment by fjmolinas** [comment]
Squashed, re-tested:

<details><summary><b> make -C tests/pkg_edhoc_c all test-with-config</b></summary>

```
main(): This is RIOT! (Version: 2021.07-devel-288-gbbe3d-pr_edhoc_c)
[initiator]: load ephemeral key: ONLY FOR TESTING
[initiator]: preset cid: ONLY FOR TESTING
[responder]: load ephemeral key: ONLY FOR TESTING
[responder]: preset cid: ONLY FOR TESTING
Starting the shell
>
>
> ifconfig
ifconfig
Iface  5  HWaddr: 1A:81:A8:28:B7:62
          L2-PDU:1500  MTU:1500  HL:64  RTR
          RTR_ADV
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::1881:a8ff:fe28:b762  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff28:b762

> init handshake fe80::1881:a8ff:fe28:b762 5683
init handshake fe80::1881:a8ff:fe28:b762 5683
[initiator]: sending msg1 (37 bytes):
0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
0xe2 0x44 0x1a 0x7c 0x21
[responder]: received an EDHOC message (len 37):
0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
0xe2 0x44 0x1a 0x7c 0x21
[responder]: sending msg2 (46 bytes):
0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
0x02 0x8d 0x19 0xcf 0x3c 0x99
[initiator]: received a message (46 bytes):
0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
0x02 0x8d 0x19 0xcf 0x3c 0x99
[initiator]: sending msg3 (20 bytes):
0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
0xe0 0xa8 0x1b 0xbf
[responder]: received an EDHOC message (len 20):
0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
0xe0 0xa8 0x1b 0xbf
[responder]: finalize exchange
[responder]: handshake successfully completed
[initiator]: handshake successfully completed
[initiator]: Transcript hash 4 (32 bytes):
0x7c 0xcf 0xde 0xdc 0x2c 0x10 0xca 0x03
0x56 0xe9 0x57 0xb9 0xf6 0xa5 0x92 0xe0
0xfa 0x74 0xdb 0x2a 0xb5 0x4f 0x59 0x24
0x40 0x96 0xf9 0xa2 0xac 0x56 0xd2 0x07

> init oscore
init oscore
OSCORE secret:
0x5b 0xb2 0xae 0xe2 0x5b 0x16 0x0e 0x7c
0x6d 0x26 0x12 ### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->0xb0 0xa6 0x01 0x09 0x16

OSCORE salt:
0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d

resp oscore
> resp oscore
OSCORE secret:
0x5b 0xb2 0xae 0xe2 0x5b 0x16 0x0e 0x7c
0x6d 0x26 0x12 0xb0 0xa6 0x01 0x09 0x16

OSCORE salt:
0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d

>
```
</details>

<details><summary><b> BOARD=samr21-xpro make -C tests/pkg_edhoc_c all test-with-config</b></summary>

```
>
> ifconfig
ifconfig
Iface  5  HWaddr: 2B:07  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK

          Long HWaddr: 00:04:25:19:18:01:AB:07
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  RTR
          6LO  IPHC
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::204:2519:1801:ab07  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff01:ab07

> init handshake fe80::204:2519:1801:ab07 5683
init handshake fe80::204:2519:1801:ab07 5683
[initiator]: sending msg1 (37 bytes):
0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
0xe2 0x44 0x1a 0x7c 0x21
[responder]: received an EDHOC message (len 37):
0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
0xe2 0x44 0x1a 0x7c 0x21
[responder]: sending msg2 (46 bytes):
0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
0x02 0x8d 0x19 0xcf 0x3c 0x99
[initiator]: received a message (46 bytes):
0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
0x02 0x8d 0x19 0xcf 0x3c 0x99
[initiator]: sending msg3 (20 bytes):
0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
0xe0 0xa8 0x1b 0xbf
[responder]: received an EDHOC message (len 20):
0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
0xe0 0xa8 0x1b 0xbf
[responder]: finalize exchange
[responder]: handshake successfully completed
[initiator]: handshake successfully completed
[initiator]: Transcript hash 4 (32 bytes):
0x7c 0xcf 0xde 0xdc 0x2c 0x10 0xca 0x03
0x56 0xe9 0x57 0xb9 0xf6 0xa5 0x92 0xe0
0xfa 0x74 0xdb 0x2a 0xb5 0x4f 0x59 0x24
0x40 0x96 0xf9 0xa2 0xac 0x56 0xd2 0x07samr21-xpeo

> init oscore
init oscore
OSCORE secret:
0x5b 0xb2 0xae 0xe2 0x5b 0x16 0x0e 0x7c
0x6d 0x26 0x12 0xb0 0xa6 0x01 0x09 0x16

OSCORE salt:
0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d

> resp oscore
resp oscore
OSCORE secret:
0x5b 0xb2 0xae 0xe2 0x5b 0x16 0x0e 0x7c
0x6d 0x26 0x12 0xb0 0xa6 0x01 0x09 0x16

OSCORE salt:
0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d

>
```
</details>


<details><summary><b> BOARD=nrf52840-mdk make -C tests/pkg_edhoc_c all test-with-config</b></summary>

```
ifconfig
Iface  5  HWaddr: 4B:38  Channel: 26  NID: 0x23  PHY: O-QPSK
          Long HWaddr: AA:04:81:69:BE:FA:4B:38
           State: IDLE
          ACK_REQ  L2-PDU:102  MTU:1280  HL:64  RTR
          6LO  IPHC
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::a804:8169:befa:4b38  scope: link  VAL
init handshake fe80::a804:8169:befa:4b38 5683
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:fffa:4b38

> init handshake fe80::a804:8169:befa:4b38 5683
[initiator]: sending msg1 (37 bytes):
0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
0xe2 0x44 0x1a 0x7c 0x21
[responder]: received an EDHOC message (len 37):
0x0d 0x00 0x58 0x20 0x8d 0x3e 0xf5 0x6d
0x1b 0x75 0x0a 0x43 0x51 0xd6 0x8a 0xc2
0x50 0xa0 0xe8 0x83 0x79 0x0e 0xfc 0x80
0xa5 0x38 0xa4 0x44 0xee 0x9e 0x2b 0x57
0xe2 0x44 0x1a 0x7c 0x21
[responder]: sending msg2 (46 bytes):
0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
0x02 0x8d 0x19 0xcf 0x3c 0x99
[initiator]: received a message (46 bytes):
0x58 0x20 0x52 0xfb 0xa0 0xbd 0xc8 0xd9
0x53 0xdd 0x86 0xce 0x1a 0xb2 0xfd 0x7c
0x05 0xa4 0x65 0x8c 0x7c 0x30 0xaf 0xdb
0xfc 0x33 0x01 0x04 0x70 0x69 0x45 0x1b
0xaf 0x35 0x37 0x4a 0xa3 0xf1 0xbd 0x5d
0x02 0x8d 0x19 0xcf 0x3c 0x99
[initiator]: sending msg3 (20 bytes):
0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
0xe0 0xa8 0x1b 0xbf
[responder]: received an EDHOC message (len 20):
0x37 0x52 0xd5 0x53 0x5f 0x31 0x47 0xe8
0x5f 0x1c 0xfa 0xcd 0x9e 0x78 0xab 0xf9
0xe0 0xa8 0x1b 0xbf
[responder]: finalize exchange
[responder]: handshake successfully completed
[initiator]: handshake successfully completed
[initiator]: Transcript hash 4 (32 bytes):
0x7c 0xcf 0xde 0xdc 0x2c 0x10 0xca 0x03
0x56 0xe9 0x57 0xb9 0xf6 0xa5 0x92 0xe0
0xfa 0x74 0xdb 0x2a 0xb5 0x4f 0x59 0x24
0x40 0x96 0xf9 0xa2 0xac 0x56 0xd2 0x07
init oscore

> init oscore
OSCORE secret:
0x5b 0xb2 0xae 0xe2 0x5b 0x16 0x0e 0x7c
0x6d 0x26 0x12 0xb0 0xa6 0x01 0x09 0x16

OSCORE salt:
0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d
resp oscore

> resp oscore
OSCORE secret:
0x5b 0xb2 0xae 0xe2 0x5b 0x16 0x0e 0x7c
0x6d 0x26 0x12 0xb0 0xa6 0x01 0x09 0x16

OSCORE salt:
0x8e 0x44 0x92 0x10 0xe0 0x3b 0xc2 0x9d
```
</details>

**Pull request merged**